File: src/common/arcan_str.c
#include "arcan_str.h"

arcan_str arcan_str_fromcstr(const char* cstr)
{
	if (cstr == NULL) return (arcan_str){0};
	const char* ch = cstr;
	while (*ch != 0) ch++;
	return (arcan_str){ .ptr = (char*)cstr, .len = ch - cstr, .mut = false };
}

bool arcan_strvalid(arcan_str str)
{
	return str.len >= 0 && !(str.ptr == NULL && str.len == 0);
}

char* arcan_stridx(arcan_str str, ptrdiff_t idx)
{
	if (!arcan_strvalid(str)) return NULL;
	if (idx < 0) {
		if (-idx > str.len) return NULL;
		return str.ptr + str.len + idx;
	}
	if (idx >= str.len) return NULL;
	return str.ptr + idx;
}

int arcan_strcmp(arcan_str s1, arcan_str s2)
{
	if (s1.ptr == s2.ptr && s1.len == s2.len) return 0;
	if (s1.len < s2.len) return -1;
	if (s1.len > s2.len) return 1;
	size_t n = s1.len < s2.len ? s1.len : s2.len;
	for (size_t i=0; i<n; ++i){
		if (s1.ptr[i] != s2.ptr[i]){
			return s1.ptr[i] < s2.ptr[i] ? -1 : 1;
		}
		i++;
	}
	return 0;
}

const char* arcan_strchr(arcan_str str, char ch)
{
	for (size_t i=0; i<str.len; ++i){
		if (str.ptr[i] == ch) return &str.ptr[i];
	}
	return NULL;
}

arcan_str arcan_strcpy(arcan_str dest, arcan_str src)
{
	if (!arcan_strvalid(dest)) return (arcan_str){0};
	if (!dest.mut || dest.len != src.len) return (arcan_str){0};
	for (size_t i=0; i<dest.len; ++i) dest.ptr[i] = src.ptr[i];
	return dest;
}

arcan_str arcan_strsub(arcan_str str, size_t from, size_t to)
{
	if (to < from || str.len < to) return (arcan_str){0};
	return (arcan_str){ .ptr = str.ptr + from, .len = to - from, .mut = str.mut };
}

bool arcan_strprefix(arcan_str str, arcan_str prefix)
{
	if (str.len < prefix.len) return false;
	arcan_str sub = arcan_strsub(str, 0, prefix.len);
	return arcan_strcmp(sub, prefix) == 0;
}

bool arcan_strpostfix(arcan_str str, arcan_str postfix)
{
	if (str.len < postfix.len) return false;
	arcan_str sub = arcan_strsub(str, str.len - postfix.len, str.len);
	return arcan_strcmp(sub, postfix) == 0;
}

arcan_str arcan_strstr(arcan_str hay, arcan_str needle)
{
	if (needle.len == 0) return (arcan_str){ .ptr = hay.ptr, .len = 0, .mut = hay.mut };
	if (hay.len < needle.len) return (arcan_str){0};
	for (size_t i=0; i<hay.len; ++i){
		arcan_str sub = arcan_strsub(hay, i, i + needle.len);
		if (arcan_strcmp(sub, needle) == 0) return sub;
	}
	return (arcan_str){0};
}

bool arcan_strtok(arcan_str str, arcan_str *tok, char seperator)
{
	if (!arcan_strvalid(str)) {
		*tok = (arcan_str){0};
		return false;
	}
	if (arcan_strvalid(*tok)){
		tok->ptr += tok->len + 1;
		if (tok->ptr > str.ptr + str.len) return false;
		tok->len = str.ptr + str.len - tok->ptr;
	} else {
		*tok = str;
	}

	const char *sep = arcan_strchr(*tok, seperator);
	if (sep == NULL) return true;
	tok->len = (sep - tok->ptr);
	return true;
}

bool arcan_strglobmatch(arcan_str str, arcan_str pattern, char wildcard)
{
	if (!arcan_strvalid(str)) return false;

	arcan_str rest = str;
	arcan_str tok = (arcan_str){0};
	if (!arcan_strtok(pattern, &tok, wildcard)) return false;
	if (!arcan_strprefix(rest, tok)) return false;
	rest = arcan_strsub(str, tok.len, str.len);

	while (arcan_strtok(pattern, &tok, wildcard)) {
		arcan_str sub = arcan_strstr(rest, tok);
		if (!arcan_strvalid(sub)) return false;
		size_t offset = (sub.ptr - rest.ptr) + tok.len;
		rest = arcan_strsub(rest, offset, rest.len);
	}

	char* last_char = arcan_stridx(pattern, -1);
	return rest.len == 0 || (last_char && *last_char == wildcard);
}

File: tests/zig/arcan_str.zig
const std = @import("std");
const c = @import("arcan_str");

const expectEqual = std.testing.expectEqual;
const expectEqualSlices = std.testing.expectEqualSlices;

fn arcan_str(str: []const u8) c.arcan_str {
    return c.arcan_str{
        .ptr = @constCast(str.ptr),
        .len = str.len,
        .mut = false,
    };
}

fn arcan_str_mut(str: []u8) c.arcan_str {
    return c.arcan_str{
        .ptr = str.ptr,
        .len = str.len,
        .mut = true,
    };
}

fn arcan_str_slice(str: c.arcan_str) []const u8 {
    return str.ptr[0..str.len];
}

fn arcan_str_empty() c.arcan_str {
    return c.arcan_str{ .ptr = null, .len = 0, .mut = false };
}

test "arcan_str_fromcstr" {
    const str = "Testing";
    const s1 = c.arcan_str_fromcstr(str.ptr);
    try expectEqual(str.len, s1.len);
    try expectEqual(str.ptr, s1.ptr);

    const s2 = c.arcan_str_fromcstr(null);
    try expectEqual(false, c.arcan_strvalid(s2));

    const empty = "\x00";
    const s3 = c.arcan_str_fromcstr(empty.ptr);
    try expectEqual(0, s3.len);
    try expectEqual(empty.ptr, s3.ptr);
}

test "arcan_stridx" {
    const str = arcan_str("Hello");
    try expectEqual(str.ptr, c.arcan_stridx(str, 0));
    try expectEqual(str.ptr + 4, c.arcan_stridx(str, 4));
    try expectEqual(null, c.arcan_stridx(str, 5));
    try expectEqual(str.ptr + 4, c.arcan_stridx(str, -1));
    try expectEqual(str.ptr, c.arcan_stridx(str, -5));
    try expectEqual(null, c.arcan_stridx(str, -6));
}

test "arcan_strcmp" {
    const s1 = arcan_str("hello");
    const s2 = arcan_str("hello world");
    const s3 = arcan_str("hell!");
    try expectEqual(0, c.arcan_strcmp(s1, s1));
    try expectEqual(-1, c.arcan_strcmp(s1, s2));
    try expectEqual(1, c.arcan_strcmp(s1, s3));
    try expectEqual(-1, c.arcan_strcmp(s3, s1));

    const s2_sub = c.arcan_str{ .ptr = s2.ptr, .len = 5 };
    try expectEqual(0, c.arcan_strcmp(s1, s2_sub));
}

test "arcan_strchr" {
    const s1 = arcan_str("Some string");
    try expectEqual(s1.ptr, c.arcan_strchr(s1, 'S'));
    try expectEqual(&s1.ptr[5], c.arcan_strchr(s1, 's'));
    try expectEqual(null, c.arcan_strchr(s1, 'w'));
}

test "arcan_strcpy" {
    var destarr = "world".*;
    const src = arcan_str("hello");
    const dest = arcan_str_mut(&destarr);
    const result = c.arcan_strcpy(dest, src);

    try expectEqual(result.ptr, dest.ptr);
    try expectEqual(result.len, dest.len);
    try expectEqualSlices(u8, arcan_str_slice(src), arcan_str_slice(result));
}

test "arcan_strsub" {
    const str = arcan_str("One two three");

    const s1 = c.arcan_strsub(str, 4, 7);
    try expectEqualSlices(u8, "two", arcan_str_slice(s1));

    const s2 = c.arcan_strsub(str, 4, 4);
    try expectEqualSlices(u8, "", arcan_str_slice(s2));

    const s3 = c.arcan_strsub(str, 4, 0);
    try expectEqual(false, c.arcan_strvalid(s3));

    const s4 = c.arcan_strsub(str, 4, 100);
    try expectEqual(false, c.arcan_strvalid(s4));

    const s5 = c.arcan_strsub(str, 8, 13);
    try expectEqualSlices(u8, "three", arcan_str_slice(s5));
}

test "arcan_strprefix" {
    const str = arcan_str("one two three");
    const p1 = arcan_str("one");
    const p2 = arcan_str("two");

    try expectEqual(true, c.arcan_strprefix(str, p1));
    try expectEqual(false, c.arcan_strprefix(str, p2));
}

test "arcan_strpostfix" {
    const str = arcan_str("one two three");
    const p1 = arcan_str("three");
    const p2 = arcan_str("two");

    try expectEqual(true, c.arcan_strpostfix(str, p1));
    try expectEqual(false, c.arcan_strpostfix(str, p2));
}

test "arcan_strstr" {
    const str = arcan_str("one two three");
    const s1 = arcan_str("three");
    const s2 = arcan_str("two");
    const s3 = arcan_str("cow");

    try expectEqual(&str.ptr[8], c.arcan_strstr(str, s1).ptr);
    try expectEqual(&str.ptr[4], c.arcan_strstr(str, s2).ptr);
    try expectEqual(false, c.arcan_strvalid(c.arcan_strstr(str, s3)));
}

test "arcan_strtok" {
    const str = arcan_str(" one two three ");

    var tok = arcan_str_empty();
    try expectEqual(true, c.arcan_strtok(str, &tok, ' '));
    try expectEqualSlices(u8, "", arcan_str_slice(tok));

    try expectEqual(true, c.arcan_strtok(str, &tok, ' '));
    try expectEqualSlices(u8, "one", arcan_str_slice(tok));

    try expectEqual(true, c.arcan_strtok(str, &tok, ' '));
    try expectEqualSlices(u8, "two", arcan_str_slice(tok));

    try expectEqual(true, c.arcan_strtok(str, &tok, ' '));
    try expectEqualSlices(u8, "three", arcan_str_slice(tok));

    try expectEqual(true, c.arcan_strtok(str, &tok, ' '));
    try expectEqualSlices(u8, "", arcan_str_slice(tok));

    try expectEqual(false, c.arcan_strtok(str, &tok, ' '));
}

test "arcan_strglobmatch" {
    const str = arcan_str("One two three");

    try expectEqual(false, c.arcan_strglobmatch(arcan_str_empty(), arcan_str("*"), '*'));
    try expectEqual(false, c.arcan_strglobmatch(str, arcan_str_empty(), '*'));

    try expectEqual(true, c.arcan_strglobmatch(arcan_str(""), arcan_str(""), '*'));
    try expectEqual(true, c.arcan_strglobmatch(arcan_str(""), arcan_str("*"), '*'));
    try expectEqual(false, c.arcan_strglobmatch(arcan_str(" "), arcan_str(""), '*'));

    try expectEqual(true, c.arcan_strglobmatch(str, str, '*'));
    try expectEqual(true, c.arcan_strglobmatch(str, arcan_str("*"), '*'));
    try expectEqual(true, c.arcan_strglobmatch(str, arcan_str("%"), '%'));

    try expectEqual(true, c.arcan_strglobmatch(str, arcan_str("One*"), '*'));
    try expectEqual(true, c.arcan_strglobmatch(str, arcan_str("*two*"), '*'));
    try expectEqual(true, c.arcan_strglobmatch(str, arcan_str("*three"), '*'));
    try expectEqual(true, c.arcan_strglobmatch(str, arcan_str("One*w*three"), '*'));

    try expectEqual(false, c.arcan_strglobmatch(str, arcan_str("One"), '*'));
    try expectEqual(false, c.arcan_strglobmatch(str, arcan_str("two"), '*'));
    try expectEqual(false, c.arcan_strglobmatch(str, arcan_str("three"), '*'));
}
